Add support for Dyson Purecool 2018 Air Purifiers models TP04 and DP04

etheralm opened the pull request:
## Description:
This PR adds support for the dyson purecool 2018 air purifier models (DP04 and TP04).  

Note: Dyson introduced a new version of their API for their 2018 models, so I had to update the upstream library to add support for them. I created a PR for those changes 6 months ago, but since it still hasn't been merged, I created a fork in order to push those changes in pypi under a different package name until it gets merged. I decided to replace the original package completely with it to avoid having 2 different packages with almost the same name which could be confusing. The new package is absolutely the same as the original one + the contents of my PR.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7846

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

etheralm committed:
Add support for Dyson Purecool Air Purifiers TP04 and DP04

etheralm committed:
Add support for Dyson Purecool Air Purifiers TP04 and DP04

etheralm committed:
fix tests line length and regenerate requirements

etheralm committed:
fix tests line length and regenerate requirements

etheralm committed:
correct rest of styling issues

etheralm committed:
correct rest of styling issues

etheralm committed:
revert incorrect change to async_added_to_hass for DysonPureCoolLinkDevice

etheralm committed:
revert incorrect change to async_added_to_hass for DysonPureCoolLinkDevice

etheralm committed:
remove unused import

etheralm committed:
remove unused import

etheralm committed:
revert incorrect change to tox.ini

etheralm committed:
revert incorrect change to tox.ini

MartinHjelmare commented on homeassistant/components/sensor/dyson.py:
```diff
@@ -183,3 +204,101 @@ def state(self):
         return None
+
+
+class DysonParticulateMatter25Sensor(DysonSensor):
```
These sensors should instead be part of the air pollutants component which is being renamed to air quality component.

So make a new platform under that component.

MartinHjelmare commented on homeassistant/components/sensor/dyson.py:
```diff
@@ -183,3 +204,101 @@ def state(self):
+        return None
+
+
+class DysonParticulateMatter10Sensor(DysonSensor):
```
See above.

MartinHjelmare commented on homeassistant/components/sensor/dyson.py:
```diff
@@ -183,3 +204,101 @@ def state(self):
+        return None
+
+
+class DysonNitrogenDioxideSensor(DysonSensor):
```
See above.

MartinHjelmare commented on homeassistant/components/sensor/dyson.py:
```diff
@@ -183,3 +204,101 @@ def state(self):
+        return None
+
+
+class DysonVolatileOrganicCompoundsSensor(DysonSensor):
```
What kind of measurement is this? Does it also fit as air quality?

beveradb replied:
It is indeed an air quality measurement, same as the Nitrogen Dioxide and Particulate Matter sensors.
Unfortunately for both the Volatile Compounds and Nitrogen Dioxide values, it's hard to say what the units should be - in my experience of using this platform over the last week, I haven't seen values for either over 100 so have assumed it's intended as a %.

fosscat replied:
Dyson seems to either measure VOC and NO2 in µg/m³ or convert to that unit because their Air Quality value is a max(PM2.5, PM10, VOC, NO2), so all four scales have to be normalized. The scale that you'll find in the Dyson Link app (0-3, 4-6, 7-8, 9+) looks like 1/10 of the api values.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -15,42 +15,71 @@
+        flow_direction_front = service.data.get(ATTR_FLOW_DIRECTION_FRONT)
+        timer = service.data.get(ATTR_TIMER)
         fan_device = next([fan for fan in hass.data[DYSON_FAN_DEVICES] if
                            fan.entity_id == entity_id].__iter__(), None)
```
Make this a generator expression instead by using parenthesis instead of brackets, and remove the `__iter__()` call.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -15,42 +15,71 @@
+SERVICE_SET_TIMER = 'dyson_set_timer'
 
 DYSON_SET_NIGHT_MODE_SCHEMA = vol.Schema({
     vol.Required(CONF_ENTITY_ID): cv.entity_id,
```
Service schema keys should use `ATTR_` constants not `CONF_`.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -15,42 +15,71 @@
+})
+
+DYSON_SET_ANGLE_SCHEMA = vol.Schema({
+    vol.Required(CONF_ENTITY_ID): cv.entity_id,
```
See above.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -15,42 +15,71 @@
+})
+
+DYSON_SET_FLOW_DIRECTION_FRONT_SCHEMA = vol.Schema({
+    vol.Required(CONF_ENTITY_ID): cv.entity_id,
```
See above.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -15,42 +15,71 @@
+})
+
+DYSON_SET_TIMER_SCHEMA = vol.Schema({
+    vol.Required(CONF_ENTITY_ID): cv.entity_id,
```
See above.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -194,13 +244,13 @@ def night_mode(self, night_mode: bool) -> None:
         return self._device.state.fan_mode == "AUTO"
 
-    def auto_mode(self, auto_mode: bool) -> None:
+    def set_auto_mode(self, auto_mode: bool) -> None:
```
Where is this used?

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -238,6 +288,210 @@ def supported_features(self) -> int:
+class DysonPureCoolDevice(FanEntity):
+    """Representation of a Dyson Purecool (TP04/DP04) fan."""
+
+    def __init__(self, hass, device):
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -238,6 +288,210 @@ def supported_features(self) -> int:
+
+    def __init__(self, hass, device):
+        """Initialize the fan."""
+        _LOGGER.debug("Creating device %s", device.name)
```
We don't want side effects like logging in init method. Please remove or move this.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -238,6 +288,210 @@ def supported_features(self) -> int:
+    def __init__(self, hass, device):
+        """Initialize the fan."""
+        _LOGGER.debug("Creating device %s", device.name)
+        self.hass = hass
```
Remove this.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -238,6 +288,210 @@ def supported_features(self) -> int:
+            self._device.disable_oscillation()
+
+    def set_night_mode(self, night_mode: bool) -> None:
+        """Turn on/off oscillating."""
```
Stale docstring.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -238,6 +288,210 @@ def supported_features(self) -> int:
+            self._device.disable_night_mode()
+
+    def set_auto_mode(self, auto_mode: bool) -> None:
+        """Turn on/off oscillating."""
```
Stale docstring.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -238,6 +288,210 @@ def supported_features(self) -> int:
+        else:
+            self._device.disable_night_mode()
+
+    def set_auto_mode(self, auto_mode: bool) -> None:
```
Where is this used?

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -238,6 +288,210 @@ def supported_features(self) -> int:
+
+    def set_flow_direction_front(self,
+                                 flow_direction_front: bool) -> None:
+        """Turn on/off oscillating."""
```
Stale docstring.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -238,6 +288,210 @@ def supported_features(self) -> int:
+
+        if self._device.state:
+            if self._device.state.speed == FanSpeed.FAN_SPEED_AUTO.value:
+                return self._device.state.speed
```
Only fan base component speed values are allowed to be returned here.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -238,6 +288,210 @@ def supported_features(self) -> int:
+        _LOGGER.debug("Turn on fan %s", self.name)
+
+        if speed is not None:
+            if speed == FanSpeed.FAN_SPEED_AUTO.value:
```
Only fan base component speed values are allowed as user interface. They need to be mapped to the dyson speed values.

MartinHjelmare commented on homeassistant/components/fan/dyson.py:
```diff
@@ -238,6 +288,210 @@ def supported_features(self) -> int:
+        from libpurecool.const import FanSpeed
+
+        supported_speeds = [
+            FanSpeed.FAN_SPEED_AUTO.value,
```
Only fan base component speed values are allowed.

MartinHjelmare commented on tests/components/fan/test_dyson.py:
```diff
@@ -325,9 +379,309 @@ def test_service_set_night_mode(self):
+    def test_dyson_purecool_turn_on(self):
+        """Test turn on purecool fan."""
+        device = _get_purecool_device_on()
+        component = dyson.DysonPureCoolDevice(self.hass, device)
```
This is not a component, it's an entity.

Please don't test the entity in isolation. Set up the component and let that set up the corresponding platform, which adds the entity. Then assert that the core state machine has the correct state for the entity.

Do this while patching appropriately to avoid I/O.

MartinHjelmare reviewed (changes requested):
The required fan speed changes will be a breaking change.

etheralm commented:
quick question, before I dive deeper in your comments. From a quick scan I can mostly say that all your comments apply to the existing dyson fan component that was there (also the tests) which I used as a base to implement the new one. Should I refactor that one too?

MartinHjelmare commented:
Yes, please refactor those too, either here or in a separate PR that we merge first.

emontnemery committed:
Update device registry of MQTT binary_sensor

emontnemery committed:
Update device registry of MQTT climate

balloob committed:
Bumped version to 0.86.0b0

balloob committed:
Sensibo to use HA operation modes (#20180)

balloob committed:
Distribute reconnect (#20181)

balloob committed:
Add command to refresh auth (#20183)

amelchio committed:
Improve Sonos discovery (#20196)

OttoWinter committed:
Bump aioesphomeapi to 1.4.2 (#20247)

* Bump aioesphomeapi to 1.4.2

* Update requirements_all.txt

amelchio committed:
Remove double logging of automation action (#20264)

balloob committed:
Updated frontend to 20190120.0

amelchio committed:
Fix 'all' entity_id in service call extraction (#20281)

amelchio committed:
Allow 'all' entity_id in service schema (#20278)

pnbruckner committed:
Handle non-string values in JSON renderer (#20233)

Handle the case of async_render_with_possible_json_value's value argument
being something other than a string. This can happen, e.g., when using the
SQL sensor to extract a datetime column such as last_changed and also using
its value_template to convert that datetime to another format. This was
causing a TypeError from json.loads, but async_render_with_possible_json_value
was only catching ValueError's.

balloob committed:
Bumped version to 0.86.0b1

balloob committed:
Clean up build artifacts correctly

balloob committed:
Updated frontend to 20190121.0

kellerza committed:
Align valid_entity_id with new slugify (#20231)

* slug

* ensure a dot

* fix

* schema_with_slug_keys

* lint

* test

balloob committed:
Bumped version to 0.86.0b2

kellerza committed:
Config Validator: schema_with_slug_keys (#20298)

* schema_with_slug_keys

* Update config_validation.py

* Update config_validation.py

balloob committed:
Updated frontend to 20190121.1

reefab committed:
Bugfix: prevent error notification when octoprint server auto detected but no configuration present. (#20303)

balloob committed:
Bumped version to 0.86.0b3

rolfberkenbosch committed:
Update locationsharinglib to version 3.0.11 (#20322)

* Update google_maps.py

There are known bug in locationsharinglib version 3.0.9, the developer has fixed this in 3.0.11. (https://github.com/costastf/locationsharinglib/issues/42)

* Update requirements_all.txt

mitchellrj committed:
Should require the 'GATTOOL' setup extras which includes pexpect. (#20263)

* Should require the 'GATTOOL' setup extras which includes pexpect.

* Also fix skybeacon's requirement and requirements_all.

balloob committed:
Fix invalid entity ID in entity registry (#20328)

syssi committed:
Fix xiaomi speed attribute name clash (#20312)

balloob committed:
Version bump to 0.86.0

fronzbot committed:
Hotfix for blink initialization failure. Fixes #20335 (#20351)

balloob committed:
Merge pull request #20354 from home-assistant/rc

0.86.0

eliseomartelli committed:
[FIX] Time reporting incorrect in sensor.gtt (#20362)

* quick fix

* remove print statement

* fixes

* remove lambda

* added pylint disable

* should be fine now

balloob committed:
Fix restore state crashing invalid entity ID (#20367)

balloob committed:
Calling save before load would crash Lovelace storage (#20368)

dgomes committed:
Fix error when API doesn't return a forecast. (#20365)

* add guard

* wrong logic

balloob committed:
Bumped version to 0.86.1

balloob committed:
Merge pull request #20369 from home-assistant/rc

0.86.1

starkillerOG committed:
Philips Hue - Remove unnessesary warning (#20394)

for white hue bulbs and bulbs from other brands like Ikea and Innr, this warning will be issued while this is not really a problem.
So just remove the warning.

balloob committed:
history allowed to load states with invalid entity IDs (#20399)

balloob committed:
Bumped version to 0.86.2

balloob committed:
Merge pull request #20402 from home-assistant/rc

0.86.2

etheralm committed:
refactor fan component and unit tests.

etheralm committed:
refactor fan component and unit tests.

etheralm committed:
fix pylint issue

etheralm committed:
fix pylint issue

LouisMT committed:
Fix HTTP login attempts check triggering too late (#20431)

sqldiablo committed:
WeMo - Fix device discovery issues (#20446)

adrum committed:
Add alarm type workaround zwave lock Yale YRD240 (#20438)

amelchio committed:
Error handling for recorder purge (#20424)

dmulcahey committed:
ZHA component rewrite part 1  (#20456)

* rearrange files

* add init to module

* update imports

* update coveragerc

* put blank line back... git raw view be damned

fabaff committed:
Upgrade astral to 1.8 (#20459)

JonasPed committed:
Correct minor comments from PR#20138. (#20454)

dgomes committed:
Utility meter (#19718)

* initial commit

* test service calls

* lint

* float -> Decimal

* extra tests

* lint

* lint

* lint

* lint

* fix self reset

* clean

* add services

* improve service example description

* add optional paused initialization

* fix

* travis fix

* fix YEARLY

* add tests for previous bug

* address comments and suggestions from @ottowinter

* lint

* remove debug

* add discoverability capabilities

* no need for _hass

* Update homeassistant/components/sensor/utility_meter.py

Co-Authored-By: dgomes <diogogomes@gmail.com>

* Update homeassistant/components/sensor/utility_meter.py

Co-Authored-By: dgomes <diogogomes@gmail.com>

* correct comment

* improve error handling

* address @MartinHjelmare comments

* address @MartinHjelmare comments

* one patch is enought

* follow @ballob suggestion in https://github.com/home-assistant/architecture/issues/131

* fix tests

* review fixes

* major refactor

* lint

* lint

* address comments by @MartinHjelmare

* rename variable

danielsjf committed:
Add Co2signal sensor (#19204)

* Initial commit for the co2signal sensor

* Clean code

* Run script gen_requirements_all.py

* remove unintended character

* Remove redundancy

* Remove unused imports

* Code style

* Code style fixes

* Code style

* Fix comments PR

Comments by @fabaff
* Remove redundant comments and variables
* Follow the latest home-assistant guidelines

* Bump CO2Signal version

* Round API result

* Improve default latitude/longitude handling

* Improve friendly name

* Improve config handling

* Make lines shorter

* Style

* Convert default to variable

None does not pass cv.string

* Message if not inclusive

* Shorten line

* Update requirements

* Update co2signal.py

Group imports;  remove empty lines; refactor use of location_type; remove logging messages; remove unused functions; add global variables

* Update co2signal.py

Import platform schema from sensor

* Small fix

* Update co2signal.py

Remove last mentions of location_type

* Review changes

Add attribution
Formatting

* Missing whitespace

* Update co2signal.py

Fix pylint

* Update co2signal.py

Change blank lines

* Update co2signal.py

Initialise _data

dmulcahey committed:
introduce gateway (#20460)

emontnemery committed:
Update device registry of MQTT fan (#20445)

balloob committed:
Update device registry of MQTT climate (#20444)

balloob committed:
Merge pull request #20442 from emontnemery/mqtt_binary_sensor_update_device_info

Update device registry of MQTT binary_sensor

emontnemery committed:
Update device registry of MQTT sensor (#20440)

emontnemery committed:
Update device registry of MQTT alarm (#20439)

emontnemery committed:
Update device registry of MQTT cover (#20443)

* Update device registry of MQTT cover

* Move config_entry to constructor

starkillerOG committed:
Philips Hue, include debug message for color gamut (#20455)

starkillerOG committed:
Philips Hue, include debug message for color gamut (#20455)

fabaff committed:
Upgrade sqlalchemy to 1.2.16 (#20474)

fabaff committed:
Upgrade psutil to 5.5.0 (#20462)

balloob committed:
Warn for old slugs/entity ids (#20478)

* Warn for old slugs/entity ids

* add comments

* Lint

* LInt

* Lint

* Lint

balloob committed:
Warn for old slugs/entity ids (#20478)

* Warn for old slugs/entity ids

* add comments

* Lint

* LInt

* Lint

* Lint

shaan7 committed:
Include exception details in the error log (#20461)

* Include exception details in the error log

I see this error quite often in my HA logs and this will be helpful for anyone who is attempting to debug this.

* Minor change

* Remove line break

balloob committed:
Bumped version to 0.86.3

balloob committed:
Merge pull request #20485 from home-assistant/rc

0.86.3

balloob committed:
Bumped version to 0.87.0.dev0

rohankapoorcom committed:
Fix linting error (#20488)

bachya committed:
Use HASS latitude/longitude as defaults for Lyft (#20491)

rohankapoorcom committed:
Upgrade zm-py to 0.3.1 (#20489)

rohankapoorcom committed:
Split out dovado to a component and sensor platform (#20339)

* Split out dovado to a component and sensor platform

* Lint

* Address code review comments (#20339)

* Switch to using a notify platform for dovado SMS (#20339)

* Optimizing imports

* Remove return on `setup_platform`.

* Clean up unneeded constants

Jc2k committed:
Add homekit_controller tests (#20457)

* Add a test for a homekit_controller switch

* Add a test for a homekit_controller lightbulb

* Add a test for homekit_controller thermostat

* Changes from review

* Patch utcnow to known time in HK tests

* Neater fixture use per review

StevenLooman committed:
Upgrade to async-upnp-client==0.14.3 (#20502)

Danielhiversen committed:
Upgrade tibber library (#20504)

* Upgrade tibber library

* Upgrade tibber library

emontnemery committed:
Update device registry of MQTT Lock (#20501)

emontnemery committed:
Update device registry of MQTT Vacuum (#20500)

emontnemery committed:
Small cleanup of MQTT platforms (#20503)

* Move CONF_UNIQUE_ID to init

* Sort imports

* Update ordering

Danielhiversen committed:
Upgrade mill library (#20514)

hfurubotten committed:
Add nilu air_quality platform (#19674)

* Add nilu air_pollutants platform

* Code Review - validation, DRYs, rm state override, new attr
- Repeated code moved to own method.
- Removed override of state property.
- New attr for showing nilu pollution recommendations.
- More validation of stations input.
- Minor fixes and typos.

* Removed unused prop

* Check for none result from client before entity add

* Moved platform to air_quality component

* Updated outdated docstrings

* Minor changes

eliseomartelli committed:
Add Iliad Italy (Mobile Telephony Provider) Sensor (#19645)

* working state

* Attrs

* > requirements generated

* linting

* fixes

* coveragerc

* ordered imports and fixed auth

* Added Throttle

* moved throttle to decorator

* remove scan interval

* lower throttle

* moved to scan interval

* Add attribution

rohankapoorcom committed:
Load/unload geofency entities correctly between component and platform (#20497)

* Load/unload geofency entities correctly between component and platform

* Lint

* Await the coroutine directly

rohankapoorcom committed:
Load/unload locative entities correctly between component and platform (#20498)

* Embed device_tracker in locative

* Load/unload locative entities correctly between component and platform

* Await the coroutine directly

* Await the correct coroutine

rohankapoorcom committed:
Load/unload gpslogger entities correctly between component and platform (#20448)

* Embed device_tracker in gpslogger

* Load/unload gpslogger entities correctly between component and platform

* Await the coroutine directly

davidlie committed:
Use more up-to-date version of pyfoscam library (#20419)

* Change foscam python library to pyfoscam, which is more up to date and has several critical bug fixes.

* Update requirements_all.txt to match.

* Inserting automatically generated requirements.txt

chrillux committed:
Reset Brottsplatskartan incident types every day (#20117)

* Reset the incident types count every day

* Remove functionality that was never implemented

We don't need to keep track of previous incidents because it's not used
anywhere.

* Create empty dictionary with a pair of braces: {}

Jc2k committed:
Add more HomeKit controller tests (#20515)

* homekit_controller tests: automatically find entity ids in tests

Some entities use dynamic ids because of the nature of the test fakes it is
hard to predict the name of the entity that will be created. This inspects the
EntityComponent of the domain to find the freshly created entity.

* homekit_controller: Tests can now define their own Service models.

All existing tests use models as defined upstream. But upstream only defines a
few service models. This adds a generic model helper for creating test
service/characteristic models.

* homekit_controller: Add cover tests

* homekit_controller: Add lock tests

* homekit_controller: Add alarm_control_panel tests

* homekit_controller: Update light tests for color_temp.

* Revert "homekit_controller tests: automatically find entity ids in tests"

This reverts commit 506caa4c3e0814eec637145d7d6eaf2bd642e99e.

* homekit_controller: Mock entity name so entity_id is consistent.

Also remove spurious subclass overrides that are identical to parent class.

* homekit_controler: Make tests less awkward as allowed top level imports

etheralm committed:
fix pylint issue, properly

etheralm committed:
Merge remote-tracking branch 'origin/dyson_new_fans' into dyson_new_fans

# Conflicts:
#	homeassistant/components/fan/dyson.py
#	requirements_all.txt

rohankapoorcom commented:
It looks like you ended up merging dev into this branch. This has the unfortunate affect of making a huge mess of the pull request. I would recommend closing out this pull request and making a new one with only your changes (rebased on top of dev).

etheralm commented:
@rohankapoorcom thanks, I noticed this late yesterday, second time I do that :). anyway I'll close this one and open a new one when all the changes that were requested are ready.
